docs/enhancements/auto-return-agent-to-infra-env.md

```diff
@@ -0,0 +1,187 @@
+---
+title: auto-return-agent-to-infra-env
+authors:
+  - "@carbonin"
+creation-date: 2022-05-17
+last-updated: 2022-06-02
+---
+
+# Automatically Return Agents to InfraEnv
+
+## Summary
+
+Without a BareMetalHost there is currently no way to automate the process
+of unbinding a host from a cluster. When a user wants to reuse a host for a new
+cluster they have to use the assisted-service unbind API, then manually boot
+the host back into the discovery image.
+
+To support automating this procedure, the assisted-service will configure nodes
+in installed clusters to run an agent. When appropriate this agent will
+reconfigure the host to boot into a discovery image, allowing the host to
+be made available for installation without the user manually interacting with
+a management system to reboot the host into the discovery image.
+
+## Motivation
+
+A user creating and managing clusters using the assisted-service may not have
+access to the systems used to configure and boot hosts. This is especially
+true in bare-metal use cases when a BareMetalHost is not present. The
+assisted-service can make this use case easier by handling the reprovisioning
+process.
+
+### Goals
+
+* Reconfigure a host to boot the discovery image as a part of unbinding from a
+  cluster without a BareMetalHost
+
+### Non-Goals
+
+* Automate graceful removal of a node from an existing cluster
+* Booting a node that is unresponsive or not currently part of an installed
+  cluster
+* Registering an agent to an InfraEnv other than the one it is currently
+  associated with
+
+## Proposal
+
+When a host should be returned to the InfraEnv use the kubeconfig created when
+the cluster was installed to run an agent on the host as a systemd service 
+using a MachineConfig. This agent will handle downloading the required 
+artifacts and rebooting the host into the discovery process.
+
+This process will only be used in situations where the assisted-service is sure
+to have connectivity to the installed cluster's API endpoint. In cases such as
+the SaaS this feature will not be used.
+
+### Timeline
+
+1. User unbinds the host from the cluster
+2. Assisted service creates a MachineConfig to run the 
+   assisted-installer-agent on the host as a systemd service
+  * If there is no BareMetalHost associated with the agent CR, the service
+    executes a new host state transition (`TransitionTypeReclaimHost`)
+  * Host moves to state `HostStatusReclaiming`
+3. The agent downloads the InfraEnv's kernel and initrd (including discovery 
+   ignition) to the host's `/boot` partition
+4. The agent creates a GRUB boot entry to boot into the discovery kernel and 
+   initrd
+5. The agent reboots the host into the discovery image
+6. Discovery agent registers host
+  * Service moves host to `HostStatusDiscoveringUnbound`
+
+### User Stories
```

Comment:
Story 2: "As a cluster creator, I want data wiped from hosts when they are removed from my clusters to avoid data leakage to other people who create and use clusters in the same environment."

If we're providing a complete host-recycling workflow, this is an important piece.

Reply:
Added